agent/grpc-external/services/resource/write.go

```diff
@@ -0,0 +1,168 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/lib/retry"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (s *Server) Write(ctx context.Context, req *pbresource.WriteRequest) (*pbresource.WriteResponse, error) {
```

Comment:
Can you add a short doc string for the function or is that better done in the proto file? I imagine quite a few controller authors will be calling/relying on this function and will want to know expectations for inputs/outputs/errors.

Reply:
Ah yes, I've been meaning to loop back and write some proper API documentation in the proto file. I think that's the best place because it'll be used as the godoc comments for the interface.